c-dependencies/js-compute-runtime/builtins/object-store.cpp

```diff
@@ -0,0 +1,464 @@
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "?");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, '*') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "*");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, '[') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "[");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, ']') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "]");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, '\n') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "newline");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  if (strchr(key_chars, '\r') != NULL) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_OBJECT_STORE_KEY_INVALID_CHARACTER, "carriage return");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+  auto acme_challenge = ".well-known/acme-challenge/";
+  if (strncmp(key_chars, acme_challenge, strlen(acme_challenge)) == 0) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_KEY_ACME);
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  if (strcmp(key_chars, ".") == 0 || strcmp(key_chars, "..") == 0) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr, JSMSG_OBJECT_STORE_KEY_RELATIVE);
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  BodyHandle body_handle = {INVALID_HANDLE};
+  int status =
+      fastly_object_store_lookup(object_store_handle(self), key_chars, key_len, &body_handle);
+  if (!HANDLE_RESULT(cx, status)) {
+    return false;
+  }
+
+  // If the handle is invalid it means no entry was found
+  // When no entry is found, we are going to resolve the Promise with `null`.
+  if (body_handle.handle == INVALID_HANDLE) {
+    JS::RootedValue result(cx);
+    result.setNull();
+    JS::ResolvePromise(cx, result_promise, result);
+  } else {
+    JS::RootedObject entry(cx, ObjectStoreEntry::create(cx, body_handle));
+    if (!entry) {
+      return false;
+    }
+    JS::RootedValue result(cx);
+    result.setObject(*entry);
+    JS::ResolvePromise(cx, result_promise, result);
+  }
+
+  args.rval().setObject(*result_promise);
+  return true;
+}
+
+bool put(JSContext *cx, unsigned argc, JS::Value *vp) {
+  METHOD_HEADER(2)
+
+  JS::RootedObject result_promise(cx, JS::NewPromiseObject(cx, nullptr));
```

Comment:
There's a lot of duplicated validation logic between `put` and `lookup`, what do you think about extracting a function do do that part?

Reply:
definitely, this commit is the one which has refactored the logic -> [d8328dd83871719549492a7af9920f0914fa28c1](https://github.com/fastly/js-compute-runtime/pull/110/commits/d8328dd83871719549492a7af9920f0914fa28c1)